zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java

```diff
@@ -1204,6 +1208,106 @@ public void testPublicPrivateNewNote() throws IOException, SchedulerException {
     ZeppelinConfiguration.create();
   }
   
+  @Test
+  public void testPersistOnRun() throws IOException, SchedulerException {
+    System.setProperty(ConfVars.ZEPPELIN_NOTEBOOK_STORAGE.getVarName(),
+        "org.apache.zeppelin.notebook.repo.GitNotebookRepo");
+    System.setProperty(ConfVars.ZEPPELIN_NOTEBOOK_PERSISTENCE.getVarName(),
+        "run");
+    ZeppelinConfiguration conf = ZeppelinConfiguration.create();
+    
+    SearchService search = mock(SearchService.class);
+    NotebookRepoSync repoManager = new NotebookRepoSync(conf);
+    Notebook notebook = new Notebook(conf, repoManager, schedulerFactory, factory,
+        interpreterSettingManager, this, search, notebookAuthorization, credentials);
+
+    // one git versioned storage initialized
+    assertThat(repoManager.getRepoCount()).isEqualTo(1);
+    
+    // persist settings applied
+    List<NotebookRepoWithSettings> settings = repoManager.getNotebookRepos(anonymous);
+    assertThat(settings.size()).isEqualTo(2);
+    assertThat(settings.get(0).name).isEqualTo("GitNotebookRepo");
+    assertThat(settings.get(1).name).isEqualTo(repoManager.getGlobalSettingsName());
+    assertThat(settings.get(1).settings.get(0).name)
+        .isEqualTo(NotebookRepoSettingUtils.NOTE_PERSISTENCE_NAME);
+    assertThat(settings.get(1).settings.get(0).selected).isEqualTo("run");
+
+    // no notes in memory
+    assertThat(notebook.getAllNotes().size()).isEqualTo(0);
+    // no notes on disk
+    assertThat(repoManager.list(anonymous).size()).isEqualTo(0);
+    
+    // create note
+    Note note = notebook.createNote(anonymous);
+    interpreterSettingManager.setInterpreters("user", note.getId(), interpreterSettingManager.getDefaultInterpreterSettingList());
+    
+    // 1 note in memory
+    assertThat(notebook.getAllNotes().size()).isEqualTo(1);
+    // no notes on disk
+    assertThat(repoManager.list(anonymous).size()).isEqualTo(0);
+    
+    // p1
+    Paragraph p1 = note.addNewParagraph(AuthenticationInfo.ANONYMOUS);
+    Map config1 = p1.getConfig();
+    config1.put("enabled", true);
+    p1.setConfig(config1);
+    p1.setText("p1");
+
+    // p2
+    Paragraph p2 = note.addNewParagraph(AuthenticationInfo.ANONYMOUS);
+    Map config2 = p2.getConfig();
+    config2.put("enabled", false);
+    p2.setConfig(config2);
+    p2.setText("p2");
+
+    // p3
+    Paragraph p3 = note.addNewParagraph(AuthenticationInfo.ANONYMOUS);
+    p3.setText("p3");
+
+    // before run
+    // 1 note with 3 paragraphs in memory
+    assertThat(notebook.getAllNotes().size()).isEqualTo(1);
+    assertNotNull(notebook.getNote(note.getId()).getParagraphs());
+    assertThat(notebook.getNote(note.getId()).getParagraphs().size()).isEqualTo(3);
+    // no notes on disk
+    assertThat(repoManager.list(anonymous).size()).isEqualTo(0);
+    
+    // when
+    note.runAll();
+
+    // wait for finish
+    while(p3.isTerminated() == false || p3.getResult() == null) {
```

Comment:
Ideally, persist could happen in a separated thread when paragraph is running, so that persist won't block the execution. Anyway,  this could be done in a separated PR as another performance improvement.